File: Largest Rectangle in Histogram - Leetcode 84/Largest Rectangle in Histogram - Leetcode 84.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length;
        Stack<int[]> stk = new Stack<>();
        int maxArea = 0;

        for (int i = 0; i < n; i++) {
            int height = heights[i];
            int start = i;
            while (!stk.isEmpty() && height < stk.peek()[0]) {
                int[] popped = stk.pop();
                int h = popped[0];
                int j = popped[1];
                int w = i - j;
                maxArea = Math.max(maxArea, h * w);
                start = j;
            }
            stk.push(new int[]{height, start});
        }

        while (!stk.isEmpty()) {
            int[] popped = stk.pop();
            int h = popped[0];
            int j = popped[1];
            int w = n - j;
            maxArea = Math.max(maxArea, h * w);
        }

        return maxArea;
    }
}

File: Largest Rectangle in Histogram - Leetcode 84/Largest Rectangle in Histogram - Leetcode 84.js
var largestRectangleArea = function(heights) {
    var n = heights.length;
    var stk = [];
    var maxArea = 0;

    for (var i = 0; i < n; i++) {
        var height = heights[i];
        var start = i;
        while (stk.length > 0 && height < stk[stk.length - 1][0]) {
            var [h, j] = stk.pop();
            var w = i - j;
            var a = h * w;
            maxArea = Math.max(maxArea, a);
            start = j;
        }
        stk.push([height, start]);
    }

    while (stk.length > 0) {
        var [h, j] = stk.pop();
        var w = n - j;
        maxArea = Math.max(maxArea, h * w);
    }

    return maxArea;
};
